src/magentic/function_call.py

```diff
@@ -47,3 +67,47 @@ def arguments(self) -> dict[str, Any]:
         signature = inspect.signature(self._function)
         bound_args = signature.bind(*self._args, **self._kwargs)
         return bound_args.arguments.copy()
+
+
+class ParallelFunctionCall(Generic[T]):
+    """A collection of FunctionCalls that can be made concurrently."""
+
+    def __init__(self, function_calls: Iterable[FunctionCall[T]]):
+        self._function_calls = CachedIterable(function_calls)
+
+    def __call__(self) -> tuple[T, ...]:
+        return tuple(function_call() for function_call in self._function_calls)
+
+    def __iter__(self) -> Iterator[FunctionCall[T]]:
+        yield from self._function_calls
+
+
+class AsyncParallelFunctionCall(Generic[T]):
+    """Async version of `ParallelFunctionCall`."""
+
+    def __init__(self, function_calls: AsyncIterable[FunctionCall[Awaitable[T] | T]]):
+        self._function_calls = CachedAsyncIterable(function_calls)
+
+    async def __call__(self) -> Tuple[T, ...]:
+        tasks_and_results: list[asyncio.Task[T] | T] = []
+        async for function_call in self._function_calls:
+            result = function_call()
+            if inspect.iscoroutine(result):
+                tasks_and_results.append(asyncio.create_task(result))
+            else:
+                result = cast(T, result)
+                tasks_and_results.append(result)
+
+        tasks = [task for task in tasks_and_results if isinstance(task, asyncio.Task)]
+        await asyncio.gather(*tasks)
+        results = [
+            task_or_result.result()
+            if isinstance(task_or_result, asyncio.Task)
+            else task_or_result
+            for task_or_result in tasks_and_results
+        ]
+        return tuple(results)
```

Comment:
Silly question: If we're returning only the results to the function calls in the `__call__`, how are you planning on recovering the `tool_call_id` [in the request back to the API](https://platform.openai.com/docs/guides/function-calling/parallel-function-calling)? Looping through the `self._function_calls`?

Reply:
Turns out the IDs don't need to be consistent across requests, just within the same request. This means we can ignore them here and generate new IDs when making a new request to openai. That happens here
https://github.com/jackmpcollins/magentic/blob/e749a9bf111bae7634e6fbca1e03971097a8283c/src/magentic/chat_model/openai_chat_model.py#L102-L114

We do now have to keep an ID on `FunctionCall` and pass that to `FunctionResultMessage` so these will have the same ID. Which changes the usage of `FunctionResultMessage` to require passing the same instance of `FunctionCall`.
https://github.com/jackmpcollins/magentic/blob/e749a9bf111bae7634e6fbca1e03971097a8283c/tests/test_chatprompt.py#L174-L186

Reply:
Ah gotcha -- the API is stateless, so we just need to make sure it stays consistent across calls.